### ISSUE TITLE
CustomScriptExtension upgrade deleted `/var/lib/waagent/custom-script`

### ISSUE THREAD

**Issue by liamdawson** [Azure/custom-script-extension-linux#114]
After an automated upgrade ran on a VM run on Azure, the Custom Script Extension stopped working. It seems that in the upgrade process, 2.0.3 was uninstalled, it deleted the `custom-script` data directory, and that when version 2.0.4 was installed, the directory was not recreated, and thus provisioning fails, as it can't work with the `custom-script/seqnum` file.

`/var/log/azure/custom-script/handler.log`:

```
+ /var/lib/waagent/Microsoft.Azure.Extensions.CustomScript-2.0.4/bin/custom-script-extension update
time=2017-11-08T21:54:44Z version=v2.0.4/git@52352b2-clean operation=update seq=59 event=start
time=2017-11-08T21:54:44Z version=v2.0.4/git@52352b2-clean operation=update seq=59 event=noop
time=2017-11-08T21:54:44Z version=v2.0.4/git@52352b2-clean operation=update seq=59 event=end
+ /var/lib/waagent/Microsoft.Azure.Extensions.CustomScript-2.0.3/bin/custom-script-extension uninstall
time=2017-11-08T21:54:45Z version=v2.0.3/git@f21ef81-clean operation=uninstall seq=59 event=start
time=2017-11-08T21:54:45Z version=v2.0.3/git@f21ef81-clean operation=uninstall seq=59 status="not reported for operation (by design)"
time=2017-11-08T21:54:45Z version=v2.0.3/git@f21ef81-clean operation=uninstall seq=59 path=/var/lib/waagent/custom-script event="removing data dir" path=/var/lib/waagent/custom-script
time=2017-11-08T21:54:45Z version=v2.0.3/git@f21ef81-clean operation=uninstall seq=59 path=/var/lib/waagent/custom-script event="removed data dir"
time=2017-11-08T21:54:45Z version=v2.0.3/git@f21ef81-clean operation=uninstall seq=59 path=/var/lib/waagent/custom-script event=uninstalled
time=2017-11-08T21:54:45Z version=v2.0.3/git@f21ef81-clean operation=uninstall seq=59 status="not reported for operation (by design)"
time=2017-11-08T21:54:45Z version=v2.0.3/git@f21ef81-clean operation=uninstall seq=59 event=end
Not writing a placeholder status file, already exists: /var/lib/waagent/Microsoft.Azure.Extensions.CustomScript-2.0.4/status/59.status
+ nohup /var/lib/waagent/Microsoft.Azure.Extensions.CustomScript-2.0.4/bin/custom-script-extension enable
time=2017-11-08T21:54:46Z version=v2.0.4/git@52352b2-clean operation=enable seq=59 event=start
time=2017-11-08T21:54:46Z version=v2.0.4/git@52352b2-clean operation=enable seq=59 event=pre-check
time=2017-11-08T21:54:46Z version=v2.0.4/git@52352b2-clean operation=enable seq=59 event="comparing seqnum" path=/var/lib/waagent/custom-script/seqnum
time=2017-11-08T21:54:46Z version=v2.0.4/git@52352b2-clean operation=enable seq=59 event="pre-check failed" error="failed to process seqnum: failed to save the sequence number: seqnum: failed to write: open /var/lib/waagent/custom-script/seqnum: no such file or directory"
Writing a placeholder status file indicating progress before forking: /var/lib/waagent/Microsoft.Azure.Extensions.CustomScript-2.0.4/status/60.status
+ nohup /var/lib/waagent/Microsoft.Azure.Extensions.CustomScript-2.0.4/bin/custom-script-extension enable
time=2017-11-09T03:14:39Z version=v2.0.4/git@52352b2-clean operation=enable seq=60 event=start
time=2017-11-09T03:14:39Z version=v2.0.4/git@52352b2-clean operation=enable seq=60 event=pre-check
time=2017-11-09T03:14:39Z version=v2.0.4/git@52352b2-clean operation=enable seq=60 event="comparing seqnum" path=/var/lib/waagent/custom-script/seqnum
time=2017-11-09T03:14:39Z version=v2.0.4/git@52352b2-clean operation=enable seq=60 event="pre-check failed" error="failed to process seqnum: failed to save the sequence number: seqnum: failed to write: open /var/lib/waagent/custom-script/seqnum: no such file or directory"
Not writing a placeholder status file, already exists: /var/lib/waagent/Microsoft.Azure.Extensions.CustomScript-2.0.4/status/60.status
+ nohup /var/lib/waagent/Microsoft.Azure.Extensions.CustomScript-2.0.4/bin/custom-script-extension enable
time=2017-11-09T03:49:24Z version=v2.0.4/git@52352b2-clean operation=enable seq=60 event=start
time=2017-11-09T03:49:24Z version=v2.0.4/git@52352b2-clean operation=enable seq=60 event=pre-check
time=2017-11-09T03:49:24Z version=v2.0.4/git@52352b2-clean operation=enable seq=60 event="comparing seqnum" path=/var/lib/waagent/custom-script/seqnum
time=2017-11-09T03:49:24Z version=v2.0.4/git@52352b2-clean operation=enable seq=60 event="pre-check failed" error="failed to process seqnum: failed to save the sequence number: seqnum: failed to write: open /var/lib/waagent/custom-script/seqnum: no such file or directory"
```

`/var/log/waagent.log`:
```
2017/11/08 21:38:41.591893 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/08 21:54:42.959541 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] Target handler state: enabled
2017/11/08 21:54:42.986050 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] [Enable] current handler state is: notinstalled
2017/11/08 21:54:43.127662 INFO Event: name=Microsoft.Azure.Extensions.CustomScript, op=Download, message=Download succeeded, duration=0
2017/11/08 21:54:43.154193 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] Initialize extension directory
2017/11/08 21:54:43.176096 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] Update settings file: 59.settings
2017/11/08 21:54:43.196726 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] Disable extension [bin/custom-script-shim disable]
2017/11/08 21:54:44.223696 INFO Event: name=Microsoft.Azure.Extensions.CustomScript, op=Disable, message=Launch command succeeded: bin/custom-script-shim disable, duration=0
2017/11/08 21:54:44.256886 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] Copy status files from old plugin to new
2017/11/08 21:54:44.287717 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] Update extension [bin/custom-script-shim update]
2017/11/08 21:54:45.314518 INFO Event: name=Microsoft.Azure.Extensions.CustomScript, op=Update, message=Launch command succeeded: bin/custom-script-shim update, duration=0
2017/11/08 21:54:45.354089 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] Uninstall extension [bin/custom-script-shim uninstall]
2017/11/08 21:54:46.384574 INFO Event: name=Microsoft.Azure.Extensions.CustomScript, op=UnInstall, message=Launch command succeeded: bin/custom-script-shim uninstall, duration=0
2017/11/08 21:54:46.421354 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] Remove extension handler directory: /var/lib/waagent/Microsoft.Azure.Extensions.CustomScript-2.0.3
2017/11/08 21:54:46.458237 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] UpdateWithInstall not set. Skip install during upgrade.
2017/11/08 21:54:46.484195 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.3] Enable extension [bin/custom-script-shim enable]
2017/11/08 21:54:47.511508 INFO Event: name=Microsoft.Azure.Extensions.CustomScript, op=Enable, message=Launch command succeeded: bin/custom-script-shim enable, duration=0
2017/11/08 21:54:50.577269 INFO [HTTP Delay] Delay 15 seconds for Status Code 503
2017/11/08 21:54:50.597550 INFO [HTTP Retry] Attempt 2 of 3: [HTTP Retry] HTTP GET Status Code 503
2017/11/08 21:55:08.717707 INFO [HTTP Delay] Delay 15 seconds for Status Code 503
2017/11/08 21:55:08.740059 INFO [HTTP Retry] Attempt 2 of 3: [HTTP Retry] HTTP GET Status Code 503
2017/11/08 22:08:43.023491 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/08 22:38:45.390917 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/08 23:08:46.702311 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/08 23:38:47.985717 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/09 00:08:50.275722 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/09 00:38:51.693627 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/09 01:08:52.869687 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/09 01:38:55.168904 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/09 02:08:56.465635 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/09 02:38:57.789849 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/09 03:09:00.221272 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/09 03:14:39.633681 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.4] Target handler state: enabled
2017/11/09 03:14:39.655277 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.4] [Enable] current handler state is: enabled
2017/11/09 03:14:39.677181 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.4] Update settings file: 60.settings
2017/11/09 03:14:39.697887 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.4] Enable extension [bin/custom-script-shim enable]
2017/11/09 03:14:40.725085 INFO Event: name=Microsoft.Azure.Extensions.CustomScript, op=Enable, message=Launch command succeeded: bin/custom-script-shim enable, duration=0
2017/11/09 03:14:40.774854 INFO Event: name=WALinuxAgent, op=ProcessGoalState, message=, duration=4304
2017/11/09 03:39:01.419783 INFO Event: name=WALinuxAgent, op=HeartBeat, message=, duration=0
2017/11/09 03:49:24.744552 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.4] Target handler state: enabled
2017/11/09 03:49:24.767538 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.4] [Enable] current handler state is: enabled
2017/11/09 03:49:24.790455 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.4] Update settings file: 60.settings
2017/11/09 03:49:24.813650 INFO [Microsoft.Azure.Extensions.CustomScript-2.0.4] Enable extension [bin/custom-script-shim enable]
2017/11/09 03:49:25.841551 INFO Event: name=Microsoft.Azure.Extensions.CustomScript, op=Enable, message=Launch command succeeded: bin/custom-script-shim enable, duration=0
2017/11/09 03:49:25.889279 INFO Event: name=WALinuxAgent, op=ProcessGoalState, message=, duration=4221
```

**Comment by liamdawson** [comment]
(I've been able to pin to version 2.0.3 of the extension, for now, which avoids the issue)

**Comment by boumenot** [comment]
Thank you for opening and issue, and I am sorry you got hit by this.  I think I have identified the issue, but I am still trying to figure out why this was not caught in our telemetry.

What region(s) are you deployed in?

**Comment by liamdawson** [comment]
In this instance, SEA.

**Comment by boumenot** [comment]
I am re-opening this issue because the fix has not been shipped far enough to replace 2.0.4.

**Comment by boumenot** [comment]
v2.0.6 has been released to address this issue.
